File: src-tauri/src/commands/system.rs
// 系统集成模块：在 VSCode 打开路径、探测本机 AI 工具版本与状态
use super::util::{command_with_path, expand_home};
use serde::{Deserialize, Serialize};
use std::process::Command;

// 单个 AI 工具的探测结果，供前端展示安装状态
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ToolStatus {
    // 工具标识，如 "claude" / "codex"
    pub id: String,
    // 展示名称
    pub name: String,
    // 是否在 PATH 中检测到该 CLI
    pub installed: bool,
    // 探测到的版本文本，未安装时为空
    pub version: String,
    // CLI 可执行文件路径，未安装时为空
    pub path: String,
}

// 工具最新版本查询结果，供前端判断本机 CLI 是否可更新
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ToolLatestVersion {
    // tool_id 存储工具标识，如 claude / codex
    pub tool_id: String,
    // package_name 存储用于查询最新版的 npm 包名
    pub package_name: String,
    // latest_version 存储 npm registry 返回的 latest 版本号
    pub latest_version: String,
}

// 在 VSCode 中打开指定文件或目录
// vscode_path 为 code CLI 路径（默认 "code"），target 为要打开的文件/目录
#[tauri::command]
pub fn open_in_vscode(vscode_path: String, target: String) -> Result<(), String> {
    // abs 为展开后的目标绝对路径
    let abs = expand_home(&target);
    // bin 为 code CLI，可执行名或绝对路径
    let bin = if vscode_path.trim().is_empty() {
        "code".to_string()
    } else {
        vscode_path
    };
    // 用修正 PATH 的命令调用 code CLI，保证 GUI 启动时也能定位到 code
    command_with_path(&bin)
        .arg(abs.to_string_lossy().to_string())
        .spawn()
        .map_err(|e| format!("调用 VSCode 失败（请确认 '{}' 可用）: {}", bin, e))?;
    Ok(())
}

// 在系统文件管理器（macOS Finder）中显示指定路径
#[tauri::command]
pub fn reveal_in_finder(target: String) -> Result<(), String> {
    // abs 为展开后的绝对路径
    let abs = expand_home(&target);
    // macOS 用 open -R 在 Finder 中定位文件
    Command::new("open")
        .arg("-R")
        .arg(abs.to_string_lossy().to_string())
        .spawn()
        .map_err(|e| format!("打开 Finder 失败: {}", e))?;
    Ok(())
}

// 探测单个 CLI 工具的版本：执行 `<bin> --version`
fn probe_tool(id: &str, name: &str, bin: &str) -> ToolStatus {
    // which 用于解析 CLI 在 PATH 中的绝对路径（用修正 PATH 保证 GUI 启动时一致）
    let resolved = command_with_path("which").arg(bin).output();
    // path 为解析出的可执行路径，未找到为空
    let path = match resolved {
        Ok(out) if out.status.success() => String::from_utf8_lossy(&out.stdout).trim().to_string(),
        _ => String::new(),
    };
    // 未找到可执行文件，直接返回未安装状态
    if path.is_empty() {
        return ToolStatus {
            id: id.to_string(),
            name: name.to_string(),
            installed: false,
            version: String::new(),
            path: String::new(),
        };
    }
    // version 通过 --version 获取，失败时留空但仍标记已安装
    let version = match command_with_path(bin).arg("--version").output() {
        Ok(out) => String::from_utf8_lossy(&out.stdout).trim().to_string(),
        Err(_) => String::new(),
    };
    ToolStatus {
        id: id.to_string(),
        name: name.to_string(),
        installed: true,
        version,
        path,
    }
}

// 探测本机已安装的 AI 工具（Claude Code CLI、Codex CLI）
#[tauri::command]
pub fn detect_tools() -> Result<Vec<ToolStatus>, String> {
    // 返回固定两个工具的探测结果，顺序稳定便于前端展示
    Ok(vec![
        probe_tool("claude", "Claude Code", "claude"),
        probe_tool("codex", "Codex CLI", "codex"),
    ])
}

// 根据工具标识返回对应 npm 包名。
// tool_id 为前端传入的工具标识，目前支持 claude / codex。
fn npm_package_for_tool(tool_id: &str) -> Option<&'static str> {
    match tool_id {
        "claude" => Some("@anthropic-ai/claude-code"),
        "codex" => Some("@openai/codex"),
        _ => None,
    }
}

// 解析 npm view 输出中的版本号。
// stdout 为 npm view <package> version 的标准输出字节。
fn parse_latest_version_stdout(stdout: &[u8]) -> Result<String, String> {
    // version 存储去掉换行与空白后的版本文本。
    let version = String::from_utf8_lossy(stdout).trim().to_string();
    if version.is_empty() {
        return Err("npm 未返回版本号".to_string());
    }

    Ok(version)
}

// 构造 npm 全局更新工具 CLI 的参数。
// package_name 为要安装到全局环境的 npm 包名。
fn build_update_tool_args(package_name: &str) -> Vec<&str> {
    vec![
        "install",
        "-g",
        package_name,
        "--registry=https://registry.npmjs.org",
    ]
}

// 查询指定工具在 npm registry 上的最新版本。
// tool_id 为工具标识，前端用该结果和本机版本比较后展示是否可更新。
#[tauri::command]
pub fn check_tool_latest_version(tool_id: String) -> Result<ToolLatestVersion, String> {
    // package_name 存储工具对应的 npm 包名。
    let package_name = npm_package_for_tool(&tool_id)
        .ok_or_else(|| format!("不支持查询 {} 的最新版本", tool_id))?;
    // output 存储 npm view 命令执行结果，registry 固定为官方源避免本机镜像滞后。
    let output = command_with_path("npm")
        .arg("view")
        .arg(package_name)
        .arg("version")
        .arg("--registry=https://registry.npmjs.org")
        .output()
        .map_err(|error| format!("查询 {} 最新版本失败: {}", package_name, error))?;

    if !output.status.success() {
        // stderr 存储 npm 查询失败时的错误文本。
        let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
        return Err(if stderr.is_empty() {
            format!("查询 {} 最新版本失败", package_name)
        } else {
            stderr
        });
    }

    // latest_version 存储解析后的最新版本号。
    let latest_version = parse_latest_version_stdout(&output.stdout)?;
    Ok(ToolLatestVersion {
        tool_id,
        package_name: package_name.to_string(),
        latest_version,
    })
}

// 更新指定工具 CLI 到 npm registry 最新版本。
// tool_id 为工具标识，后端会映射到对应 npm 包并执行全局安装。
#[tauri::command]
pub fn update_tool_cli(tool_id: String) -> Result<String, String> {
    // package_name 存储工具对应的 npm 包名。
    let package_name =
        npm_package_for_tool(&tool_id).ok_or_else(|| format!("不支持更新 {} 的 CLI", tool_id))?;
    // args 存储 npm install -g 的完整参数。
    let args = build_update_tool_args(package_name);
    // output 存储 npm 全局安装命令的执行结果。
    let output = command_with_path("npm")
        .args(args)
        .output()
        .map_err(|error| format!("更新 {} 失败: {}", package_name, error))?;

    if !output.status.success() {
        // stderr 存储 npm 更新失败时的错误文本。
        let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
        return Err(if stderr.is_empty() {
            format!("更新 {} 失败", package_name)
        } else {
            stderr
        });
    }

    // stdout 存储 npm 更新成功时的输出文本。
    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    Ok(if stdout.is_empty() {
        format!("{} 已更新到最新版本", package_name)
    } else {
        stdout
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_npm_package_for_tool_maps_claude_code() {
        assert_eq!(
            npm_package_for_tool("claude").unwrap(),
            "@anthropic-ai/claude-code"
        );
    }

    #[test]
    fn test_parse_latest_version_stdout_trims_output() {
        assert_eq!(
            parse_latest_version_stdout(b"2.1.196\n").unwrap(),
            "2.1.196"
        );
    }

    #[test]
    fn test_build_update_tool_args_uses_global_install_and_official_registry() {
        assert_eq!(
            build_update_tool_args("@anthropic-ai/claude-code"),
            vec![
                "install",
                "-g",
                "@anthropic-ai/claude-code",
                "--registry=https://registry.npmjs.org"
            ]
        );
    }
}

File: src/core/system.js
// 系统集成：打开 VSCode/Finder、探测 AI CLI、查询和更新 CLI 最新版本。
import { spawn } from "node:child_process";
import { expandHome, buildCommandEnv, runCommand } from "./util.js";

// OFFICIAL_NPM_REGISTRY 存储查询和安装 AI CLI 时使用的官方 npm registry。
const OFFICIAL_NPM_REGISTRY = "https://registry.npmjs.org";

// npmPackageForTool 根据工具标识返回 npm 包名。
// toolId 参数存储工具标识，如 claude / codex。
export function npmPackageForTool(toolId) {
  if (toolId === "claude") {
    return "@anthropic-ai/claude-code";
  }
  if (toolId === "codex") {
    return "@openai/codex";
  }
  return undefined;
}

// releaseNotesUrlForTool 返回指定工具的官方更新内容网址。
// toolId 参数存储工具标识，如 claude / codex。
export function releaseNotesUrlForTool(toolId) {
  // releaseNotesUrls 存储工具标识到官方 changelog 或 releases 页面的映射。
  const releaseNotesUrls = {
    claude: "https://github.com/anthropics/claude-code/blob/main/CHANGELOG.md",
    codex: "https://github.com/openai/codex/releases",
  };

  return releaseNotesUrls[toolId] ?? "";
}

// binForTool 根据工具标识返回 CLI 可执行文件名。
// toolId 参数存储工具标识，如 claude / codex。
export function binForTool(toolId) {
  if (toolId === "claude") {
    return "claude";
  }
  if (toolId === "codex") {
    return "codex";
  }
  return undefined;
}

// parseLatestVersionStdout 解析 npm view 输出中的版本号。
// stdout 参数存储 npm 输出文本。
export function parseLatestVersionStdout(stdout) {
  // version 存储去掉换行与空白后的版本文本。
  const version = String(stdout || "").trim();
  if (!version) {
    throw new Error("npm 未返回版本号");
  }
  return version;
}

// parseToolVersionStdout 从 CLI --version 输出中提取 semver 版本号。
// stdout 参数存储 CLI 版本输出文本。
export function parseToolVersionStdout(stdout) {
  // match 存储版本输出里可用于和 npm latest 对比的 semver 片段。
  const match = String(stdout || "").match(/\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?/);
  return match?.[0] || "";
}

// buildPackageSpec 构造带版本或 tag 的 npm 包规格。
// packageName 参数存储 npm 包名，version 参数存储目标版本或 tag。
export function buildPackageSpec(packageName, version = "latest") {
  // targetVersion 存储规范化后的目标版本；空值兜底到 latest。
  const targetVersion = String(version || "latest").trim() || "latest";
  return `${packageName}@${targetVersion}`;
}

// buildUpdateToolArgs 构造 npm 全局更新工具 CLI 的参数。
// packageName 参数存储要安装到全局环境的 npm 包名，version 参数存储目标版本或 tag。
export function buildUpdateToolArgs(packageName, version = "latest") {
  return [
    "install",
    "-g",
    buildPackageSpec(packageName, version),
    `--registry=${OFFICIAL_NPM_REGISTRY}`,
  ];
}

// buildVoltaUpdateToolArgs 构造 Volta 重新安装工具 CLI 的参数。
// packageName 参数存储要交给 Volta 管理的 npm 包名，version 参数存储目标版本或 tag。
export function buildVoltaUpdateToolArgs(packageName, version = "latest") {
  return ["install", buildPackageSpec(packageName, version)];
}

// installManagerForToolPath 根据当前命中的 CLI 路径判断应使用的安装管理器。
// toolPath 参数存储 which 命令解析出的可执行文件路径。
export function installManagerForToolPath(toolPath) {
  // normalizedPath 存储规范化为字符串后的 CLI 路径。
  const normalizedPath = String(toolPath || "");
  if (normalizedPath.includes("/.volta/bin/")) {
    return "volta";
  }
  return "npm";
}

// spawnAndForget 启动外部命令，不等待命令完成。
// bin 参数存储可执行文件名，args 参数存储命令参数。
function spawnAndForget(bin, args) {
  // child 存储子进程句柄。
  const child = spawn(bin, args, {
    env: buildCommandEnv(),
    stdio: "ignore",
    detached: true,
  });
  child.unref();
}

// openInVscode 在 VSCode 中打开指定文件或目录。
// vscodePath 参数存储 code CLI 路径，target 参数存储目标路径。
export function openInVscode(vscodePath, target) {
  // bin 存储最终使用的 code 命令。
  const bin = String(vscodePath || "").trim() || "code";
  // absoluteTarget 存储展开后的目标路径。
  const absoluteTarget = expandHome(target);
  spawnAndForget(bin, [absoluteTarget]);
}

// revealInFinder 在 macOS Finder 中显示指定路径。
// target 参数存储目标文件或目录。
export function revealInFinder(target) {
  // absoluteTarget 存储展开后的目标路径。
  const absoluteTarget = expandHome(target);
  spawnAndForget("open", ["-R", absoluteTarget]);
}

// probeTool 探测单个 CLI 工具的版本与路径。
// id/name/bin 参数分别存储工具标识、展示名和可执行文件名。
async function probeTool(id, name, bin) {
  try {
    // whichOutput 存储 which 命令输出。
    const whichOutput = await runCommand("which", [bin]);
    // path 存储解析出的可执行路径。
    const path = whichOutput.stdout.trim();
    if (!path) {
      throw new Error("not found");
    }

    // versionOutput 存储 --version 输出。
    const versionOutput = await runCommand(bin, ["--version"]).catch(() => ({ stdout: "" }));
    return {
      id,
      name,
      installed: true,
      version: versionOutput.stdout.trim(),
      path,
    };
  } catch {
    return {
      id,
      name,
      installed: false,
      version: "",
      path: "",
    };
  }
}

// detectTools 探测本机已安装的 Claude Code CLI 与 Codex CLI。
export async function detectTools() {
  return Promise.all([
    probeTool("claude", "Claude Code", "claude"),
    probeTool("codex", "Codex CLI", "codex"),
  ]);
}

// checkToolLatestVersion 查询指定工具在 npm registry 上的最新版本。
// toolId 参数存储工具标识，如 claude / codex。
export async function checkToolLatestVersion(toolId) {
  // packageName 存储工具对应的 npm 包名。
  const packageName = npmPackageForTool(toolId);
  if (!packageName) {
    throw new Error(`不支持查询 ${toolId} 的最新版本`);
  }

  // output 存储 npm view 命令输出。
  const output = await runCommand("npm", [
    "view",
    packageName,
    "version",
    `--registry=${OFFICIAL_NPM_REGISTRY}`,
  ]);
  return {
    tool_id: toolId,
    package_name: packageName,
    latest_version: parseLatestVersionStdout(output.stdout),
    release_notes_url: releaseNotesUrlForTool(toolId),
  };
}

// resolveToolPath 查询当前 PATH 中实际命中的 CLI 路径。
// bin 参数存储 CLI 可执行文件名。
async function resolveToolPath(bin) {
  // whichOutput 存储 which 命令输出。
  const whichOutput = await runCommand("which", [bin]);
  // toolPath 存储去除空白后的实际 CLI 路径。
  const toolPath = whichOutput.stdout.trim();
  if (!toolPath) {
    throw new Error(`未找到 ${bin} 可执行文件`);
  }
  return toolPath;
}

// verifyToolVersion 确认更新后当前 CLI 真实版本已经等于目标版本。
// bin 参数存储 CLI 名称，packageName 参数存储 npm 包名，targetVersion 参数存储目标版本，toolPath 参数存储更新前命中的路径。
async function verifyToolVersion(bin, packageName, targetVersion, toolPath) {
  // versionOutput 存储 CLI 更新后的 --version 输出。
  const versionOutput = await runCommand(bin, ["--version"]);
  // installedVersion 存储从 CLI 输出中提取出的真实版本。
  const installedVersion = parseToolVersionStdout(versionOutput.stdout);
  if (installedVersion === targetVersion) {
    return;
  }

  // reportedVersion 存储用于错误提示的版本文本，解析失败时保留原始输出。
  const reportedVersion = installedVersion || versionOutput.stdout.trim() || "未知";
  // pathHint 存储当前 PATH 命中的二进制提示，帮助定位 Volta/npm/Homebrew 多来源冲突。
  const pathHint = toolPath ? `当前命中的可执行文件：${toolPath}` : "未能解析当前可执行文件路径";
  throw new Error(
    `已执行更新命令，但 ${bin} 仍报告版本 ${reportedVersion}，目标版本是 ${targetVersion}。${pathHint}。请检查 PATH 中是否还有其它 ${packageName} 安装来源。`,
  );
}

// updateToolCli 更新指定工具 CLI 到 npm registry 最新版本。
// toolId 参数存储工具标识，如 claude / codex。
export async function updateToolCli(toolId) {
  // packageName 存储工具对应的 npm 包名。
  const packageName = npmPackageForTool(toolId);
  if (!packageName) {
    throw new Error(`不支持更新 ${toolId} 的 CLI`);
  }

  // bin 存储工具对应的 CLI 可执行文件名。
  const bin = binForTool(toolId);
  if (!bin) {
    throw new Error(`不支持更新 ${toolId} 的 CLI`);
  }

  // latestInfo 存储 npm registry 返回的最新版本信息，用精确版本避免安装到旧 tag 或缓存版本。
  const latestInfo = await checkToolLatestVersion(toolId);
  // targetVersion 存储本次更新要安装的精确版本。
  const targetVersion = latestInfo.latest_version;
  // toolPath 存储当前 PATH 中实际命中的 CLI 路径，用于选择安装管理器。
  const toolPath = await resolveToolPath(bin);
  // installManager 存储根据当前 CLI 路径判断出的安装管理器。
  const installManager = installManagerForToolPath(toolPath);
  // output 存储安装命令输出。
  const output =
    installManager === "volta"
      ? await runCommand("volta", buildVoltaUpdateToolArgs(packageName, targetVersion), {
          env: { npm_config_registry: OFFICIAL_NPM_REGISTRY },
        })
      : await runCommand("npm", buildUpdateToolArgs(packageName, targetVersion));

  await verifyToolVersion(bin, packageName, targetVersion, toolPath);

  // stdout 存储成功时的输出文本。
  const stdout = output.stdout.trim();
  return stdout || `${packageName} 已更新到 ${targetVersion}`;
}
